cpp/tests/experimental/indexing/point_quadtree_test.cu

```diff
@@ -0,0 +1,221 @@
+#include <cuspatial/vec_2d.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <thrust/copy.h>
+#include <thrust/device_vector.h>
+#include <thrust/fill.h>
+
+namespace {
+
+template <typename T>
+thrust::host_vector<T> inline device_uvector_to_host(rmm::device_uvector<T> const& uvec)
+{
+  thrust::host_vector<T> hvec(uvec.size());
+  cudaMemcpyAsync(hvec.data(),
+                  uvec.data(),
+                  uvec.size() * sizeof(T),
+                  cudaMemcpyKind::cudaMemcpyDeviceToHost,
+                  uvec.stream());
+  uvec.stream().synchronize();
+  return hvec;
+}
+
+}  // namespace
+
+template <typename T>
+struct QuadtreeOnPointIndexingTest : public cudf::test::BaseFixture {
+  void test(std::vector<cuspatial::vec_2d<T>> const& points,
+            const cuspatial::vec_2d<T> vertex_1,
+            const cuspatial::vec_2d<T> vertex_2,
+            const T scale,
+            const int8_t max_depth,
+            const int32_t max_size,
+            std::vector<uint32_t> const& expected_key,
+            std::vector<uint8_t> const& expected_level,
+            std::vector<bool> const& expected_is_parent_node,
+            std::vector<uint32_t> const& expected_length,
+            std::vector<uint32_t> const& expected_offset)
+  {
+    thrust::device_vector<cuspatial::vec_2d<T>> d_points{points};
+
+    auto pair = cuspatial::quadtree_on_points(
+      d_points.begin(), d_points.end(), vertex_1, vertex_2, scale, max_depth, max_size, mr());
+
+    auto& point_indices = pair.first;
+    EXPECT_EQ(point_indices.size(), points.size());
+
+    auto& tree             = pair.second;
+    auto& key_d            = tree.key;
+    auto& level_d          = tree.level;
+    auto& is_parent_node_d = tree.is_parent_node;
+    auto& length_d         = tree.length;
+    auto& offset_d         = tree.offset;
+
+    EXPECT_EQ(key_d.size(), expected_key.size());
+    EXPECT_EQ(level_d.size(), expected_level.size());
+    EXPECT_EQ(is_parent_node_d.size(), expected_is_parent_node.size());
+    EXPECT_EQ(length_d.size(), expected_length.size());
+    EXPECT_EQ(offset_d.size(), expected_offset.size());
+
+    auto key_h            = device_uvector_to_host<uint32_t>(key_d);
+    auto level_h          = device_uvector_to_host<uint8_t>(level_d);
+    auto is_parent_node_h = device_uvector_to_host<bool>(is_parent_node_d);
+    auto length_h         = device_uvector_to_host<uint32_t>(length_d);
+    auto offset_h         = device_uvector_to_host<uint32_t>(offset_d);
+
+    EXPECT_THAT(key_h, ::testing::Pointwise(::testing::Eq(), expected_key));
+    EXPECT_THAT(level_h, ::testing::Pointwise(::testing::Eq(), expected_level));
+    EXPECT_THAT(is_parent_node_h, ::testing::Pointwise(::testing::Eq(), expected_is_parent_node));
+    EXPECT_THAT(length_h, ::testing::Pointwise(::testing::Eq(), expected_length));
+    EXPECT_THAT(offset_h, ::testing::Pointwise(::testing::Eq(), expected_offset));
```

Comment:
Using the new `expect_vector_equivalent` should simplify this. Although we might need to extend it to support integers.